DSpace/DSpace
# Batch process for IIIF OCR indexing

## Description

This is an initial attempt to support IIIF Search API indexing using the [dbmdz Solr OCR Highlighting plugin](https://github.com/dbmdz/solr-ocrhighlighting/). The PR  includes a pre-configured `word_highlighting` Solr core.

A new DSpace CLI task is provided for indexing at the Community, Collection, or Item level.

Since one of the unique features of that plugin is the ability to lazy load OCR files from disk, this PR relies on a separate service written in `Go` and described here: https://github.com/mspalti/solr_ocr_processor   Binary files for the **ocr_processor** can  be copied and ran on the Solr host if needed to support lazy loading. 

The ocr_processor release includes  files for Linux, MacOS and Windows deployment. 

https://github.com/mspalti/solr_ocr_processor/releases/tag/v1.0.0


## Instructions for Reviewers

List of changes in this PR:
* A new CLI `iiif-search-index` task.
* A preconfigured Solr core for the `word_highlighing` plugin in `extras/iiif`
* New DSpace iiif configuration options for the `WordHighlightSolrSearch` plugin

There's probably a lot that needs discussion so I'm happy to consider this a _work in progress_. I'll be stress testing soon by indexing collections with tens of thousands of ALTO files. That may reveal additional work that needs doing. To date the code has been tested extensively with small collections that include a few hundred OCR files.

_The rationale for the standalone **ocr_processor** service is to support "lazy" loading of OCR files from disk when Solr is running on a different host than DSpace._ "Lazy" loading is not required by the `word_highlighting` plugin but it is a useful option. One quick way to support lazy loading (without additional magic in the data center) is to run the alto_indexer service on the same host as Solr. Compiling binaries for various platforms is easy in the `Go` language, and deployment is simple. A Docker deployment option is also provided. More details on the service are described in the ocr_processor [README](https://github.com/mspalti/alto_indexer). 

Quick description of how this works. 

OCR files are added to the Item `OtherContent` bundle so that they can be retrieved via the DSpace IIIF service. The one constraint is that the METS file must be named `mets.xml` for easy lookup. The METS file will drive subsequent retrieval and processing of OCR files and assure that processing order and identifiers align with images and canvases in the Manifest.

OCR file processing steps are managed by configuration properties found in the alto_indexer `config.yml` file.  The main options are whether to:

- convert ALTO or hOCR files to MiniOcr,
- write the OCR file to disk (adding only the file pointer to the Solr index) 
- instead add the full contents of the OCR file to the index
-  convert Unicode codepoints to XML-escaped ASCII characters when lazy loading is used.

Converting to MiniOcr is recommended.

The `iiif-search-index` CLI task traverses DSpace records at the Community, Collection, or Item levels. It executes a GET request to the alto_indexer to see if the item has already been indexed. If it has not been indexed, it sends a POST request that triggers Solr indexing for that Item. The `iiif-search-index` CLI task can also be used to delete OCR files from the Solr index and from disk if lazy loading.

**To test this service:** 

1. Retrieve the OCR processor archive file from https://github.com/mspalti/solr_ocr_processor/releases/tag/v1.0.0
2. Update the OCR processor `configs/config.yml` file (note in particular paths to the log file directory and the directory for OCR files if using lazy loading).
3. Start the service from the command line: `./ocr_processor`
4. Copy the `word_highlighting` solr core from `extras/iiif` to your Solr `configsets` directory. _If you set the `index_type` use "full" in `config.yml` you will need to edit  the Solr core's `schema.xml` and comment out the `ExternalUtf8ContentFilterFactory` filter that is used when loading OCR files from disk._
5. Load the Solr core.
6. Create a DSpace Item to index with OCR files in the `OtherContent` bundle. If using a METS file, make sure it is named `mets.xml`
7. Add `dspace.iiif.enabled` and `iiif.search.enabled` metadata to the Item.
8. Use the following CLI commands to index or delete:

_./bin/dspace iiif-search-index --add -e user@dspace.edu -i f797f6ee-f27f-4548-8590-45d6df8a7431
./bin/dspace iiif-search-index --delete -e user@dspace.edu -i f797f6ee-f27f-4548-8590-45d6df8a7431_

You should be able to see results in Solr by querying:

_http://solrhost:8983/solr/#/word_highlighting/query?q=ocr_text:query%20term&q.op=OR&indent=true&hl=true&hl.ocr.fl=ocr_text_

You should also see search results in the Mirador viewer when you query on the search term.

One more thing. If you chose to convert Unicode to XML-escaped codepoints you will need to tell the `WordHighlightSolrSearch` plugin. Do this by setting the new `iiif.search.index.xml.encode` in `config/modules/iiif.cfg` to be true.


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
